python/glow/wgr/linear_model/ridge_model.py

```diff
@@ -312,6 +312,56 @@ def transform(self,
 
         return pivoted_df
 
+    def transform_loco(self,
+                       blockdf: DataFrame,
+                       labeldf: pd.DataFrame,
+                       sample_blocks: Dict[str, List[str]],
+                       modeldf: DataFrame,
+                       cvdf: DataFrame,
+                       covdf: pd.DataFrame = pd.DataFrame({}),
+                       chromosomes: List[str] = []) -> pd.DataFrame:
+        """
+        Generates predictions for the target labels in the provided label DataFrame by applying the model resulting from
+        the RidgeRegression fit method to the starting block matrix using a leave-one-chromosome-out (LOCO) scheme.
+
+        Args:
+            blockdf : Spark DataFrame representing the beginning block matrix X
+            labeldf : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            modeldf : Spark DataFrame produced by the RidgeRegression fit method, representing the reducer model
+            cvdf : Spark DataFrame produced by the RidgeRegression fit method, containing the results of the cross
+            validation routine.
+            covdf : Pandas DataFrame containing covariates to be included in every model in the stacking
+            ensemble (optional).
+            chromosomes : List of chromosomes to leave out during the LOCO scheme (optional). If not provided, the
+            chromosomes will be inferred from the block matrix.
+
+        Returns:
+            Pandas DataFrame containing prediction y_hat values per chromosome. The rows are indexed by sample ID and
+            chromosome; the columns are indexed by label. The column types are float64. The DataFrame is sorted using
+            chromosome as the primary sort key, and sample ID as the secondary sort key.
+        """
+        # Regex captures the chromosome name in the header
+        # level 1 header: chr_3_block_8_alpha_0_label_sim100
+        # level 2 header: all_block_3_alpha_0_label_sim100
```

Comment:
If the first element is `all`, does that not mean "all chromosomes"? It's pretty weird that the chromosome does not come first in that case.

Reply:
The first term represents the header block, which is `chr_N` for the level-1 reduced block DF and `all` for the level-2 reduced block DF.

Reply:
@LelandBarnard Would it make sense to replace `all` with `chr_N` for the level-2 reduced block DF?

Reply:
Yes, this is a bit confusing.  I think after 2 levels what we really want are headers like `chr_3_alpha_0_label_sim100` or possibly `all_chr_3_alpha_0_label_sim100` but with the hacky way I had written the `new_headers` function it was expecting a pattern like `<outer_block>_<inner_block>_<alpha>_<label>`.  After one round of reduction, the outer block is the chromosome and the inner block is the header block, but after two rounds the outer block is the whole genome (hence the `all`) and the inner block is the chromosome (so `chr_3` becomes `block_3`).  Definitely open to suggestions for cleaning this up.  I think Karen has made a lot of progress already with the improvements to the `new_headers` function.

Reply:
Thanks - I took another hack at this to match your first suggestion. Could you take another look? I also hashed the chromosome term if it sets the inner index, so we can match non-numeric chromosome names (eg. X, Y).